### PR TITLE
Build with Numpy>=2.0 + no more Linux i689 + setup PDM

### LINKED ISSUE #384: Support for Numpy 2.0
Currently, pyFFTW cannot be compiled with Numpy 2.0.

I'm working in this issue with PR #383 but I don't understand the build error happening on Windows.

I see

```
INFO:root:"C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\bin\HostX86\x64\link.exe" "" /nologo /INCREMENTAL:NO /LTCG /MANIFEST:EMBED,ID=1 /LIBPATH:D:\a\pyFFTW\pyFFTW\pyfftw /LIBPATH:C:\Users\runneradmin\AppData\Local\Temp\cibw-run-xl2r4zss\cp310-win_amd64\build\venv\bin /LIBPATH:C:\Users\runneradmin\AppData\Local\Temp\cibw-run-xl2r4zss\cp310-win_amd64\build\venv\lib /LIBPATH:C:\Users\runneradmin\AppData\Local\Temp\cibw-run-xl2r4zss\cp310-win_amd64\build\venv\libs /LIBPATH:C:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.10.11\tools\libs /LIBPATH:C:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.10.11\tools /LIBPATH:C:\Users\runneradmin\AppData\Local\Temp\cibw-run-xl2r4zss\cp310-win_amd64\build\venv\PCbuild\amd64 "/LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.22621.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\\lib\10.0.22621.0\\um\x64" "/LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.22621.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\\lib\10.0.22621.0\\um\x64" libfftw3_omp-3.lib libfftw3-3.lib D:\Users\RUNNER~1\AppData\Local\Temp\pyfftw-q9xyo862\fftw_init_threads.obj /OUT:C:\Users\RUNNER~1\AppData\Local\Temp\pyfftw-q9xyo862\a.out.exe
    LINK : fatal error LNK1181: cannot open input file '.obj'
```

and then:

```
INFO:root:"C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DPYFFTW_HAVE_DOUBLE=1 -DPYFFTW_HAVE_DOUBLE_OMP=0 -DPYFFTW_HAVE_DOUBLE_THREADS=1 -DPYFFTW_HAVE_DOUBLE_MULTITHREADING=1 -DPYFFTW_HAVE_DOUBLE_MPI=0 -DPYFFTW_HAVE_SINGLE=1 -DPYFFTW_HAVE_SINGLE_OMP=0 -DPYFFTW_HAVE_SINGLE_THREADS=1 -DPYFFTW_HAVE_SINGLE_MULTITHREADING=1 -DPYFFTW_HAVE_SINGLE_MPI=0 -DPYFFTW_HAVE_LONG=1 -DPYFFTW_HAVE_LONG_OMP=0 -DPYFFTW_HAVE_LONG_THREADS=1 -DPYFFTW_HAVE_LONG_MULTITHREADING=1 -DPYFFTW_HAVE_LONG_MPI=0 -DPYFFTW_HAVE_MPI=0 -ID:\a\pyFFTW\pyFFTW\include -ID:\a\pyFFTW\pyFFTW\pyfftw -IC:\Users\runneradmin\AppData\Local\Temp\pip-build-env-9nstmgbv\overlay\Lib\site-packages\numpy\_core\include -IC:\Users\runneradmin\AppData\Local\Temp\cibw-run-xl2r4zss\cp310-win_amd64\build\venv\include -ID:\a\pyFFTW\pyFFTW\include\win -IC:\Users\runneradmin\AppData\Local\Temp\cibw-run-xl2r4zss\cp310-win_amd64\build\venv\include -IC:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.10.11\tools\include -IC:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.10.11\tools\Include -IC:\Users\runneradmin\AppData\Local\Temp\cibw-run-xl2r4zss\cp310-win_amd64\build\venv\include -IC:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.10.11\tools\include -IC:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.10.11\tools\Include "-IC:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\include" "-IC:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\ATLMFC\include" "-IC:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Auxiliary\VS\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\um" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\shared" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\winrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\cppwinrt" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" "-IC:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\include" "-IC:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.40.33807\ATLMFC\include" "-IC:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Auxiliary\VS\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\um" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\shared" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\winrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\cppwinrt" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" /Tcpyfftw\pyfftw.c /Fobuild\temp.win-amd64-cpython-310\Release\pyfftw\pyfftw.obj
    pyfftw.c
    C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-9nstmgbv\overlay\Lib\site-packages\numpy\_core\include\numpy\npy_1_7_deprecated_api.h(14) : Warning Msg: Using deprecated NumPy API, disable it with #define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
    D:\a\pyFFTW\pyFFTW\pyfftw\fftw3.h(358): error C2061: syntax error: identifier 'fftw_complex'
    D:\a\pyFFTW\pyFFTW\pyfftw\fftw3.h(358): error C2059: syntax error: ';'
    D:\a\pyFFTW\pyFFTW\pyfftw\fftw3.h(358): error C2143: syntax error: missing ')' before '*'
    D:\a\pyFFTW\pyFFTW\pyfftw\fftw3.h(358): error C2081: 'fftw_complex': name in formal parameter list illegal
    D:\a\pyFFTW\pyFFTW\pyfftw\fftw3.h(358): error C2143: syntax error: missing '{' before '*'
    D:\a\pyFFTW\pyFFTW\pyfftw\fftw3.h(358): error C2143: syntax error: missing ';' before '*'
    D:\a\pyFFTW\pyFFTW\pyfftw\fftw3.h(358): error C2059: syntax error: 'type'
    D:\a\pyFFTW\pyFFTW\pyfftw\fftw3.h(358): error C2059: syntax error: ')'
    D:\a\pyFFTW\pyFFTW\pyfftw\fftw3.h(358): fatal error C1003: error count exceeds 100; stopping compilation
```

### PR DESCRIPTION
**Pull request by paugier**
Fixes #384

Initially, this PR was focused on supporting Numpy>=2.0 but I had to include other things:

- Centos 7 has reached EOL on 1 July 2024, so manylinux2014 is broken, so I don't see how to build wheels for Linux i686. This was already possible only for Python 3.9. Since Numpy does not build such wheels, I guess it makes sense to stop supporting this platform.

- When I was struggling with a local error fixed in a7a8ca0fe81b50c7ba988e7f7e7a5bb12dc3038c I setup [PDM](https://pdm-project.org), which can be used by contributors. In my experience, it is very useful. I added few lines in the README to explain how PDM can be used to setup a developer environment. Moreover, we now have a lockfile. I added a simpler CI to test with the packages listed in this lockfile.

- I transferred some metadata from setup.py to pyproject.toml.

- I skipped few tests related to coercion of complex data since it is not supported anymore by Numpy 2.0. It's not ideal, but it was the simplest solution. I'm happy to work a bit on pyFFTW but unfortunately I can't spend too much time on it.

### CODE REVIEW

**Comment by paugier** [comment]
"Centos 7 has reached EOL today 1 July 2024", which explains the issue on Linux (nothing to do with this PR).

https://serverfault.com/questions/1161816/mirrorlist-centos-org-no-longer-resolve

**Comment by paugier** [comment]
I sent a message on Numpy mailing list on the issue on Windows...

https://mail.python.org/archives/list/numpy-discussion@python.org/thread/XTM5XEGRXSGWWQBSYIL6DYZZMTXCGEOG/

**Comment by hgomersall** [comment]
Thanks for letting me know. I have no idea what the problem is (I have not much experience building on Windows).

**Comment by mattip** [comment]
The first question is "which fftw3.h is being used here"? Searching the codebase I see this one, but line 358 is empty: https://github.com/pyFFTW/pyFFTW/blob/7dc338519a3e76d9688b268c3f82be6b43d70a05/include/win/fftw3.h#L350-L360

Let's assume that the correct line is 355. I think the problem may go back to numpy including `complex.h` in version 2.0, which I think will trip up this line https://github.com/pyFFTW/pyFFTW/blob/7dc338519a3e76d9688b268c3f82be6b43d70a05/include/win/fftw3.h#L57-L60. You may want to rearrange the include directives so that numpy is included last (or at least after `fftw3.h` when compiling `pyfftw.c`

**Comment by hgomersall** [comment]
Interesting. It might be worth using a later pyfftw3.h at this point. I'm not quite sure why it _is_ included - can it not be arranged to find the header file?

**Comment by paugier** [comment]
Thanks @mattip. I think that fftw3.h of the error is not pyFFTW/include/win/fftw3.h but is actually downloaded from https://fftw.org/pub/fftw/fftw-3.3.5-dll64.zip during the CI (https://github.com/pyFFTW/pyFFTW/blob/master/.github/workflows/wheel_tests_and_release.yml#L167C1-L173C1).

**Comment by paugier** [comment]
I found in include/pyfftw_complex.h (included in pyfftw.c):

https://github.com/pyFFTW/pyFFTW/blob/7dc338519a3e76d9688b268c3f82be6b43d70a05/include/pyfftw_complex.h#L36-L52

It seems to be incompatible with Numpy 2.0 which includes complex.h. I don't know what would be a clean solution.

**Comment by hgomersall** [comment]
Finished?

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,65 @@
...
+    "setuptools",
+    "Cython>=3",
+    # taken from scipy (2024/06/28)
+    # numpy requirement for wheel builds for distribution on PyPI - building
+    # against 2.x yields wheels that are also compatible with numpy 1.x at
+    # runtime.
+    # Note that building against numpy 1.x works fine too - users and
+    # redistributors can do this by installing the numpy version they like and
+    # disabling build isolation.
+    "numpy>=2.0.0rc1",
```
Comment (mattip) [inline pyproject.toml:13]: nit: 2.0.0 has been released...
```suggestion
    "numpy>=2.0.0",
```

File: setup.py [hunk 2]
```diff
@@ -644,6 +644,9 @@ def build_extensions(self):
         self._pyfftw_define_macros = [
             (f"PYFFTW_{k}", int(v)) for k, v in sniffer.compile_time_env.items()
         ]
+        self._pyfftw_define_macros.append(("NPY_NO_DEPRECATED_API", "NPY_1_7_API_VERSION"))
+        # because Numpy 2.0 includes complex.h
+        self._pyfftw_define_macros.append(("FFTW_NO_Complex", "1"))
```
Comment (mattip) [inline setup.py:649]: Cool!

**Review by mattip — COMMENTED** [review]
It seems to work, I wonder if `FFTW_NO_Complex` should only be declared on windows.

**Comment by paugier** [comment]
> It seems to work, I wonder if `FFTW_NO_Complex` should only be declared on windows.

I did that but to be honest I don't understand what is be better and why it works on Linux and Mac without it!

**Comment by paugier** [comment]
@hgomersall, I think it's ready to be reviewed!

**Review by hgomersall — APPROVED** [review]

**Comment by hgomersall** [comment]
Looks good. we might want to add a note that 32-bit builds are no longer supported.

**Comment by mattip** [comment]
> I don't understand what is be better and why it works on Linux and Mac without it!

Linux and Mac have full support for `_Complex` as a language feature. Windows does not even after importing `complex.h`. The stanza in question uses heuristics to determine whether there is full support or not. If not it falls back to representing complex as two floats: `float fftw_complex[2]` instead of using a language builtin. By including numpy's headers, the stanza in question thinks that the compiler properly supports `_Complex`, but MSVC does not. Luckily fftw provided an escape hatch via defining `FFTW_NO_COMPLEX` to get the old behaviour.

I hope that not only helped but is correct.

**Comment by paugier** [comment]
> Looks good. we might want to add a note that 32-bit builds are no longer supported.

I added a sentence in the README.

**Pull request merged**